Move `SourceMap` initialization

nnethercote opened the pull request:
So it happens at the same time as `SessionGlobals` initialization, rather than shortly afterward.

r? @WaffleLapkin

nnethercote committed:
Factor out common code in interface tests.

Replacing `mk_session` with `sess_and_cfgs`, which does a bit more of
the shared stuff -- the option parsing and the `build_configuration`
call.

nnethercote committed:
Construct `SourceMap` at the same time as `SessionGlobals`.

Currently `SourceMap` is constructed slightly later than
`SessionGlobals`, and inserted. This commit changes things so they are
done at the same time.

Benefits:
- `SessionGlobals::source_map` changes from
  `Lock<Option<Lrc<SourceMap>>>` to `Option<Lrc<SourceMap>>`. It's still
  optional, but mutability isn't required because it's initialized at
  construction.
- `set_source_map` is removed, simplifying `run_compiler`, which is
  good because that's a critical function and it's nice to make it
  simpler.

This requires moving things around a bit, so the necessary inputs are
available when `SessionGlobals` is created, in particular the `loader`
and `hash_kind`, which are no longer computed by `build_session`. These
inputs are captured by the new `SourceMapInputs` type, which is threaded
through various places.

nnethercote committed:
Move `initialize_checked_jobserver`.

Currently it's a method on `EarlyDiagCtxt`, which is not the right place
for it at all -- `EarlyDiagCtxt` is used to issue diagnostics, but
shouldn't be doing any of the actual checking.

This commit moves it into a standalone function that takes an
`EarlyDiagCtxt` as an argument, which is more sensible. This does
require adding `EarlyDiagCtxt::early_struct_warn`, so a warning can be
returned and then modified with a note. (And that likely explains why
somebody put `initialize_checked_jobserver` into `EarlyDiagCtxt` in the
first place.)

nnethercote commented:
This is a follow-up to #122810, and the first two commits come from that PR. I've put this in a separate PR so we can merge the removal of `CodegenBackend::target_override` first, and see if that causes any problems before making changes that build on that.

WaffleLapkin commented on compiler/rustc_interface/src/tests.rs:
```diff
@@ -25,42 +25,46 @@ use std::num::NonZero;
-    );
-    let cfg = parse_cfg(&sess.dcx(), matches.opt_strs("cfg"));
-    (sess, cfg)
+    rustc_span::create_session_globals_then(DEFAULT_EDITION, || {
```
Why not `create_default_session_globals_then`?

Also, am I reading this right that this is test only?

nnethercote replied:
It's overwritten by the subsequent commit, but I changed it anyway.

Yes, this is test-only code.

WaffleLapkin commented on compiler/rustc_span/src/lib.rs:
```diff
@@ -104,35 +104,35 @@ pub struct SessionGlobals {
+    /// The session's source map, if there is one. This field should only be
+    /// used in places where the `Session` is truly not available, such as
+    /// `<Span as Debug>::fmt`.
+    source_map: Option<Lrc<SourceMap>>,
```
Why is this an option? The previous version of the comment is obviously outdated, so there is another reason now?

nnethercote replied:
There are lots of cases (mostly in tests) where we don't bother setting a source map. That was true previously, but the old comment didn't explain that was the second reason for having `Option`.

WaffleLapkin commented on compiler/rustc_span/src/source_map.rs:
```diff
@@ -167,9 +167,17 @@ struct SourceMapFiles {
 }
 
+/// Used to construct a `SourceMap` with `SourceMap::with_inputs`.
+pub struct SourceMapInputs(
+    pub Box<dyn FileLoader + Send + Sync>,
+    pub FilePathMapping,
+    pub SourceFileHashAlgorithm,
+);
```
Would it be better to use a named struct?

nnethercote replied:
ok

WaffleLapkin reviewed (approved):
This mostly looks good. I have a couple comments, but we'll have to wait for #122811 to merge first anyway.

nnethercote commented:
I have updated and addressed the comments. I think this shouldn't land quite yet, to give a little more time for any fallout from #122810 to happen. I will be around on Monday but then I will be on vacation for three weeks. Maybe this should wait until I get back, just to be safe.

WaffleLapkin commented:
r=me whenever you feel like we waited enough :)

bors commented:
:umbrella: The latest upstream changes (presumably #115220) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
There have been no complaints about #122810. I think this is safe to merge.

@bors r=WaffleLapkin

bors commented:
:pushpin: Commit 9b0ced000a7e9db6bed97046974adfa9bf3bc165 has been approved by `WaffleLapkin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=WaffleLapkin 9b0ced000a7e9db6bed97046974adfa9bf3bc165 -->
<!-- homu: {"type":"Approved","sha":"9b0ced000a7e9db6bed97046974adfa9bf3bc165","approver":"WaffleLapkin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
